en/lessons/basics/collections.md

```diff
@@ -3,120 +3,287 @@ version: 1.3.1
+
+```haskell
+ghci> :t (&&)
+(&&) :: Bool -> Bool -> Bool
+ghci> :t uncurry (&&)
+uncurry (&&) :: (Bool, Bool) -> Bool
+ghci> :t curry (uncurry (&&))
+curry (uncurry (&&)) :: Bool -> Bool -> Bool
+```
+
+### Constructing Tuples
+
+We can construct tuples using a familiar syntax `(a,b)`.
+
+```haskell
+ghci> myTuple = (True,"hello")
+ghci> :t myTuple
+myTuple :: (Bool, [Char])
+```
+
+### Consuming Tuples
+
+We can consume tuples by breaking them apart. With a 2-tuple the options are pretty simple. They have the predictable names `fst` and `snd`, and are available in the standard library, so no need to import anything!
+
+```haskell
+ghci> fst myTuple
+True
+ghci> snd myTuple
+"hello"
+```
+
+The types of these functions simple, yet reveal exactly how they behave. `fst` picks the first element and `snd` picks the second element.
+
+```haskell
+ghci> :t fst
+fst :: (a, b) -> a
+ghci> :t snd
+snd :: (a, b) -> b
+```
+
+We can also pattern match on tuples to extract their contents.
+
+```haskell
+ghci> (\(a,b) -> not a) myTuple
+False
+ghci> (\(a,b) -> b ++ " world") myTuple
+"hello world"
+```
+
+### Limitations of Tuples
+
+The main limitation with tuples in haskell is that each different length of tuple is a distinct data type. That means there is no general function to add an additional element to any tuple. Rather these kinds of functions need to be defined ad hoc.
+
+Here is an example of increasing the tuple's length.
+
+```haskell
+ghci> twoTupleToThreeTuple c (a,b) = (a,b,c)
+ghci> twoTupleToThreeTuple () myTuple
+(1,"world",())
+```
+
+Trying to call our function on a tuple of the incorrect length results in a type error telling us just as much; the function expected a 2-tuple but we gave it a 3-tuple.
+
+```haskell
+ghci> twoTupleToThreeTuple True (1,2,3)
+
+<interactive>:19:27: error:
+    • Couldn't match expected type: (a, b)
+                  with actual type: (a0, b0, c0)
+                  ...
+```
 ## Lists
 
-Lists are simple collections of values which may include multiple types; lists may also include non-unique values:
+From a usability perspective, Lists solve the problem of extensibility that tuples face, but they can only contain one type (in other words they are homogenous). Lists also have a special built in syntax.
 
-```elixir
-iex> [3.14, :pie, "Apple"]
-[3.14, :pie, "Apple"]
+```haskell
+ghci> [1,2,3,4]
+[1,2,3,4]
+ghci> ["hello", True]
+
+<interactive>:22:11: error:
+    • Couldn't match type ‘Bool’ with ‘[Char]’
+      Expected: String
+        Actual: Bool
+    ...
+```
+
+### Inductive Types
+
+Lists are also a first introduction to an "inductive" or "recursive" type. Here is an example that is identical to the haskell implementation without the syntactic sugar (we use backticks to add infix syntax).
+
+```haskell
+data List a = Nil | a `Cons` List a
```

Comment:
Do you think that maybe this should be `Cons a (List a)` instead?

Reply:
Yeah, I was concerned about getting too deep into fixity here. I just wanted to keep the syntax similar to the actual list syntax.

Reply:
I agree: that infix and expecialy ticks may be more confusing. My vote would be to hold off such tricks till the chapter on data types.